src/wp-admin/includes/image-edit.php

```diff
@@ -1026,12 +1038,177 @@ function wp_save_image( $post_id ) {
+					}
+					// When the targeted thumbnail is selected no additional size and subsize is set.
+					continue;
+				}
+
+				$main_images[ $targeted_mime ]     = array(
+						'path' => $new_path,
+						'file' => pathinfo( $new_path, PATHINFO_BASENAME ),
+				);
+				$subsized_images[ $targeted_mime ] = $meta['sizes'];
+				continue;
+			}
+
+			if ( ! isset( $allowed_mimes[ $targeted_mime ] ) || ! is_string( $allowed_mimes[ $targeted_mime ] ) ) {
+				continue;
+			}
+
+			if ( ! $img::supports_mime_type( $targeted_mime ) ) {
+				continue;
+			}
+
+			$extension = explode( '|', $allowed_mimes[ $targeted_mime ] );
+			$extension = $extension[0];
+
+			// If the target is `thumbnail` make sure only that size is generated.
+			if ( 'thumbnail' === $target ) {
+				if ( ! isset( $subsized_images[ $post->post_mime_type ]['thumbnail']['file'] ) ) {
+					continue;
+				}
+				$thumbnail_file    = $subsized_images[ $post->post_mime_type ]['thumbnail']['file'];
+				$current_extension = pathinfo( $thumbnail_file, PATHINFO_EXTENSION );
+
+				// Create a file with then new extension out of the targeted file.
+				$target_file_name     = preg_replace( "/\.$current_extension$/", ".$extension", $thumbnail_file );
+				$target_file_location = path_join( $original_directory, $target_file_name );
+				$result               = $img->save( $target_file_location, $targeted_mime );
+
+				if ( is_wp_error( $result ) ) {
+					continue;
+				}
+
+				$subsized_images[ $targeted_mime ] = array( 'thumbnail' => $result );
+			} else {
+				$destination = trailingslashit( $original_directory ) . "{$filename}.{$extension}";
+				$result      = $img->save( $destination, $targeted_mime );
+
+				if ( is_wp_error( $result ) ) {
+					continue;
+				}
+
+				$main_images[ $targeted_mime ]     = $result;
+				$subsized_images[ $targeted_mime ] = $img->multi_resize( $resize_sizes );
+			}
+		}
+
+		foreach ( $mime_transforms as $targeted_mime ) {
+			// Make sure the path and file exists as those values are required.
+			$image_directory = null;
+			if ( isset( $main_images[ $targeted_mime ]['path'], $main_images[ $targeted_mime ]['file'] ) && file_exists( $main_images[ $targeted_mime ]['path'] ) ) {
+				// Add sources to original image metadata.
+				$meta['sources'][ $targeted_mime ] = array(
+						'file'     => $main_images[ $targeted_mime ]['file'],
+						'filesize' => wp_filesize( $main_images[ $targeted_mime ]['path'] ),
+				);
+				$image_directory                   = pathinfo( $main_images[ $targeted_mime ]['path'], PATHINFO_DIRNAME );
+			}
+
+			/**
+			 * If no original image was provided the image_directory can't be determined, in that scenario try to
+			 * find it from the `file` property.
+			 *
+			 * @see get_attached_file()
+			 */
+			if (
+					null === $image_directory
+					&& isset( $meta['file'] )
+					&& 0 !== strpos( $meta['file'], '/' )
+					&& ':\\' !== substr( $meta['file'], 1, 2 )
+			) {
```

Comment:
Can we reduce the indentation here?

```suggestion
			if (
				null === $image_directory
				&& isset( $meta['file'] )
				&& 0 !== strpos( $meta['file'], '/' )
				&& ':\\' !== substr( $meta['file'], 1, 2 )
			) {
```